### PR TITLE
feat(ppx): separate `@@deriving abstract` into `jsProperties`, `getSet`

### PR DESCRIPTION
**Pull request by anmonteiro**

### CODE REVIEW

File: ppx/melange_ppx.ml
Comment (jchavarri) [inline ppx/melange_ppx.ml:963]: is "make" necessary, or could it be just `opt_keys`? or `opt_js_keys` as you [mentioned](https://github.com/melange-re/melange/pull/979#issuecomment-1856183928) including "js" somewhere.

File: ppx/melange_ppx.ml
Comment (jchavarri) [inline ppx/melange_ppx.ml:952]: What do you think would be the plans for `abstract`? Would it be deprecated in another PR?

File: test/blackbox-tests/deriving-separate.t
Comment (jchavarri) [inline test/blackbox-tests/deriving-separate.t:63]: Just curious, why is this warning silencing needed?

File: ppx/ast_derive/ast_derive_abstract.ml
Comment (jchavarri) [inline ppx/ast_derive/ast_derive_abstract.ml:165]: This function seems to be mostly copied from `handleTdcl` above. Is there a way to reuse more? Or because the plan is to remove the other one, doesn't matter?

**Review by jchavarri — APPROVED** [review]
I think this looks great. As mentioned, it's inconsistent now to have ppx payloads with OCaml casing like the ones added here and JS casing like `jsConverter`, but it's not a big deal in any case.

File: ppx/melange_ppx.ml
Comment (anmonteiro) [reply ppx/melange_ppx.ml:963]: yeah, I'm open to changing the name; I'd prefer keeping something that mentions it makes a JS object, e.g. `jsConstructor` or `jsObject`

**Comment by anmonteiro** [comment]
> it's inconsistent now to have ppx payloads with OCaml casing like the ones added here and JS casing like `jsConverter`, but it's not a big deal in any case.

oh absolutely. I'll change that before merging for sure

File: test/blackbox-tests/deriving-separate.t
Comment (anmonteiro) [reply test/blackbox-tests/deriving-separate.t:63]: here's one example where you reported the warning existing when we generate externals: https://github.com/melange-re/melange/pull/415

File: ppx/melange_ppx.ml
Comment (anmonteiro) [reply ppx/melange_ppx.ml:952]: yeah, I suppose we should probably deprecate it if we're going ahead with this. as I mentioned in https://github.com/melange-re/melange/pull/987#discussion_r1428095102, I'll implement abstract in terms of the new functions too.

File: ppx/melange_ppx.ml
Comment (jchavarri) [reply ppx/melange_ppx.ml:963]: I was thinking about the JS [hasOwnProperty](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty) method. It seems a property is the mix of [a key and a value](https://developer.mozilla.org/en-US/docs/Glossary/Property/JavaScript), so maybe the deriver name should include it.

It'd be nice to define it like "constructor that skip properties in the JS object if the field is `None`", maybe `optJsProperties`, `makeOptProperties`? or `skipPropertiesConstructor`?

Comment (jchavarri) [reply ppx/melange_ppx.ml:963]: > I'd prefer keeping something that mentions it makes a JS object,

I am not sure this is very informative. All records compile now to JS objects so a name related to that wouldn't be very meaningful imo.

**Comment by anmonteiro** [comment]
https://github.com/melange-re/melange/pull/987/commits/9d55611f66e3857be1d813d3c26db648dcc4c7f3 implements `abstract` in terms of the 2 new ones.

Will deprecate it in a subsequent PR.

File: ppx/melange_ppx.ml
Comment (anmonteiro) [reply ppx/melange_ppx.ml]: ok, I called it `jsProperties` to unblock the PR, but we can brainstorm a better name later on.

**Pull request merged**
